libraries/chain/finality/qc.cpp

```diff
@@ -0,0 +1,503 @@
+         pending_state = state_t::weak_achieved;
+
+      if (weak_sum > max_weak_sum_before_weak_final) {
+         if (pending_state == state_t::weak_achieved)
+            pending_state = state_t::weak_final;
+         else if (pending_state == state_t::unrestricted)
+            pending_state = state_t::restricted;
+      }
+      break;
+
+   case state_t::weak_achieved:
+      if (weak_sum >= max_weak_sum_before_weak_final)
+         pending_state = state_t::weak_final;
+      break;
+
+   case state_t::weak_final:
+   case state_t::strong:
+      // getting another weak vote... nothing to do
+      break;
+   }
+   return vote_status::success;
+}
+
+// thread safe
+vote_status open_qc_sig_t::add_vote(uint32_t connection_id, block_num_type block_num,
+                                    bool strong, size_t index,
+                                    const bls_signature& sig, uint64_t weight) {
+   std::unique_lock g(*_mtx);
+   state_t pre_state = pending_state;
+   vote_status s = strong ? add_strong_vote(index, sig, weight)
+                          : add_weak_vote(index, sig, weight);
+   state_t post_state = pending_state;
+   g.unlock();
+
+   fc_dlog(vote_logger, "connection - ${c} block_num: ${bn}, vote strong: ${sv}, status: ${s}, pre-state: ${pre}, post-state: ${state}, quorum_met: ${q}",
+        ("c", connection_id)("bn", block_num)("sv", strong)("s", s)("pre", pre_state)("state", post_state)("q", is_quorum_met(post_state)));
+   return s;
+}
+
+// called by get_best_qc which acquires a mutex
+qc_sig_t open_qc_sig_t::extract_qc_sig_from_open() const {
+   qc_sig_t qc_sig;
+
+   if( pending_state == state_t::strong ) {
+      qc_sig.strong_votes = strong_votes.bitset;
+      qc_sig.sig          = strong_votes.sig;
+   } else if (is_quorum_met_no_lock()) {
+      qc_sig.strong_votes = strong_votes.bitset;
+      qc_sig.weak_votes   = weak_votes.bitset;
+      qc_sig.sig          = strong_votes.sig;
+      qc_sig.sig.aggregate(weak_votes.sig);
+   } else
+      assert(0); // this should be called only when we have an open qc with a quorum
+
+   return qc_sig;
+}
+
+std::optional<qc_sig_t> open_qc_sig_t::get_best_qc() const {
+   std::lock_guard g(*_mtx);
+   // if this does not have a valid QC, consider received_qc_sig only
+   if( !is_quorum_met_no_lock() ) {
+      if( received_qc_sig ) {
+         return std::optional{*received_qc_sig};
+      }
+      return {};
+   }
+
+   qc_sig_t qc_sig_from_open = extract_qc_sig_from_open();
+
+   // if received_qc_sig does not have value, consider valid_qc_sig_from_open only
+   if( !received_qc_sig ) {
+      return std::optional{std::move(qc_sig_from_open)};
+   }
+
+   // Both received_qc_sig and qc_sig_from_open have value. Compare them and select a better one.
+   // Strong beats weak. Tie-break by received_qc_sig, strong beats weak
+   bool use_received_qc_sig = received_qc_sig->is_strong() || (received_qc_sig->is_weak() && qc_sig_from_open.is_weak());
```

Comment:
If it is not strong, it has to be weak :-)

```suggestion
   bool use_received_qc_sig = received_qc_sig->is_strong() || qc_sig_from_open.is_weak();
```